### PR TITLE
MC-17524: Child product is not addable to Cart from PLP if it does not have default source assigned

### PR DESCRIPTION
**Pull request by namlier**
### Description (*)
Grouped Product has possibility to setup default qty of its units. It means if the number is setted such a Grouped product will be possible to add to Cart from Product Listing Page directly. In the Cart, accordingly, will appear all the grouped product items with qty as it was setted before. Though, if some of the units is out of stock, such units won't appear in product cart. This algorythm implements by OutOfStockFilter plugin.

But this plugin doesn't counts MSI extension. So that's why it is suggested to change the plugin with the new one which would be check Stock Status considering current website's stock in the MSI's implementation of GroupedProduct.

### Manual testing scenarios (*)
1. Create Grouped product with one Simple product (in stock, with positive qties)
2. Create Custom source.
3. Create Stock, link it with the Custom source and set a channel to Default Website. (therefore Default Website will work with the Custom stock and not with the Default Stock anymore)
4. In the before created Simple product set qty of default source as 0, and assing it to the Custom source with In stock status and positive qty.
5. Find the Grouped product in the Product Listing page and add it to cart using an appropriative button.

Expected result:
The Simple product will appears in the Shopping cart.

Actual result:
Success message will appears and no items will shown in the Shopping cart.

### Questions or comments
Pay attention to how i check the products is salable or not, maybe there is another to do it for whole set of items in one request to database.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/inventory?pullRequest=2386) <br/>All committers have signed the CLA.

**Comment by ishakhsuvarov** [comment]
@magento run all tests

File: app/code/Magento/InventoryGroupedProduct/composer.json
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryGroupedProduct/composer.json]: Most of the new dependencies seem to be overhead, this module is only supposed to extend Grouped Product.

File: app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php]: This plugin seems to be pretty much the same as `\Magento\GroupedCatalogInventory\Plugin\OutOfStockFilter` meaning that both of them would be now executed on each operation. This might have a noticeable effect on performance and generally is not the most optimal design.

File: app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php]: While this solution is functional, there are still at least 2 alternatives to duplicating the logic in this plugin:
* Longer solution would be to adapt `StockStatusCriteriaInterface` and `StockStatusRepositoryInterface::getList` to take advantage of MSI. This might require building additional plugins and and manipulation with queries in the Inventory modules. This solution, while could be functional, does not have too much sense, as all of this work would be done to extend a deprecated API
* Simpler and more clean solution would be refactoring original plugin (`Magento\GroupedCatalogInventory\Plugin\OutOfStockFilter`) to use `StockRegistryInterface` which is already adapted by MSI to support multi-stock capabilities.

File: app/code/Magento/InventoryGroupedProduct/etc/di.xml
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryGroupedProduct/etc/di.xml]: Almost completely duplicates existing plugin.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
There are multiple concerns regarding this solution at the moment:
* Duplication of plugins and logic would at least double the execution time, however in reality it would be even more. Existing plugin could be refactored slightly to use MSI-ready APIs.
* Set of included fixtures seems to be too large. Please carefully check existing fixtures to make sure new ones are needed.

File: app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php
Comment (namlier) [reply app/code/Magento/InventoryGroupedProduct/Plugin/OutOfStockFilter.php]: Thank you for your consideration. I have reworked this solution on CE's OutOfStockFilter plugin.
Screenshot from 2019-07-12 16-28-21
, where $this->stockRegistry is an object of StockRegistryInterface class.

Can it be a right solution?

**Comment by namlier** [comment]
> * Set of included fixtures seems to be too large. Please carefully check existing fixtures to make sure new ones are needed.
I have checked, and didn't found fixture that links one source with one stock. There is only fixtures for many websites and stocks, but my test only needs to change a stock for any website. I think i can rework the test to use many websites from the existing fixtures, but is it worth?

About dependencies in composer.json, they are required by the fixtures that are used in the test. What we can do in this case?

**Comment by namlier** [comment]
We have to wait until changes with fix from CE will be merged into this repository.

**Comment by ishakhsuvarov** [comment]
Hi @mastiuhin-olexandr 

Can you please check if the changes needed for this to function correctly are already merged? Os there anything else needed to make it work?

Thanks

**Review by sidolov — APPROVED** [review]

**Review by sidolov — APPROVED** [review]

**Pull request merged**
